cyral/resource_cyral_sidecar_listener.go

```diff
@@ -0,0 +1,363 @@
+				HttpMethod: http.MethodPost,
+				CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+					return fmt.Sprintf("https://%s/v1/sidecars/%s/listeners",
+						c.ControlPlane,
+						d.Get(SidecarIdKey).(string))
+
+				},
+				NewResourceData: func() ResourceData { return &SidecarListenerResource{} },
+				NewResponseData: func(_ *schema.ResourceData) ResponseData { return &CreateListenerAPIResponse{} },
+			}, ReadSidecarListenersConfig,
+		),
+		ReadContext: ReadResource(ReadSidecarListenersConfig),
+		UpdateContext: UpdateResource(
+			ResourceOperationConfig{
+				Name:       "SidecarListenersResourceUpdate",
+				HttpMethod: http.MethodPut,
+				CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+					return fmt.Sprintf("https://%s/v1/sidecars/%s/listeners/%s",
+						c.ControlPlane,
+						d.Get(SidecarIdKey).(string),
+						d.Get(ListenerIdKey).(string))
+
+				},
+				NewResourceData: func() ResourceData { return &SidecarListenerResource{} },
+			}, ReadSidecarListenersConfig,
+		),
+		DeleteContext: DeleteResource(
+			ResourceOperationConfig{
+				Name:       "SidecarListenersResourceDelete",
+				HttpMethod: http.MethodDelete,
+				CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+					return fmt.Sprintf("https://%s/v1/sidecars/%s/listeners/%s",
+						c.ControlPlane,
+						d.Get(SidecarIdKey).(string),
+						d.Get(ListenerIdKey).(string))
+				},
+			},
+		),
+		Schema: map[string]*schema.Schema{
+			ListenerIdKey: {
+				Description: "ID of the listener that will be bound to the sidecar.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Computed:    true,
+			},
+			SidecarIdKey: {
+				Description: "ID of the sidecar that the listener will be bound to.",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			RepoTypesKey: {
+				Description: "List of repository types that the listener supports. Currently limited to one repo type from supported repo types:" + supportedTypesMarkdown(repositoryTypes()),
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			NetworkAddressKey: {
+				Description:  "The network address that the sidecar listens on.",
+				Type:         schema.TypeSet,
+				Optional:     true, // This field is required, due to ExactlyOneOf directive
+				ExactlyOneOf: []string{NetworkAddressKey},
+				MaxItems:     1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						HostKey: {
+							Description: "Host where the sidecar will listen for the given repository. Omit to listen on all interfaces.",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						PortKey: {
+							Description: "Port where the sidecar will listen for the given repository.",
+							Type:        schema.TypeInt,
+							Required:    true,
+						},
+					},
+				},
+			},
+			MySQLSettingsKey: {
```

Comment:
I believe the API already validates that at least one of MySQLSettingsKey, S3SettingsKey or DynamoDbSettingsKey needs to be specified, but another option would be to make this a static validation in the terraform side by changing these arguments to `schema.TypeList` and use the `ExactlyOneOf`. Im usually in favor of static validation (it also helps reflect the correct validations to the auto-generated docs), but this is just a suggestion, feel free to ignore.

Reply:
What does this mean? Most of these are optional (unless listener is of type dynamo). 
mysq doesn't need one, nor does s3, or any of the others (oracle, sqlserver, pg etc)

Reply:
@gengdahlCyral you're right, the correct would be to use ConflictsWith instead of ExactlyOneOf, this way the settings would still be optional but wouldn't be able to be specified together for the same listener. [See](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Schema).